src/App.tsx

```diff
@@ -1,24 +1,219 @@
-/* eslint-disable max-len */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, {
+  useMemo,
+  useEffect,
+  useState,
+  useCallback,
+} from 'react';
+import cn from 'classnames';
+import { Todo, UpdateTodoArgs } from './types/Todo';
+import {
+  createTodo,
+  deleteTodo,
+  getTodos,
+  patchTodo,
+} from './api/todos';
 import { UserWarning } from './UserWarning';
+import { TodoList } from './components/TodoList';
+import { TodoForm } from './components/TodoForm';
+import { ErrorMessage } from './components/ErrorMessage';
+import { StatusFilterType } from './types/StatusFilterType';
+import { StatusFilter } from './components/StatusFilter';
 
-const USER_ID = 0;
+const USER_ID = 10897;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState(StatusFilterType.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([0]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => setError('Loading todos failed'));
+  }, []);
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
+
+    if (error) {
+      timeout = setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [error]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const filteredTodos = useMemo(() => {
+    switch (filter) {
+      case StatusFilterType.COMPLETED:
+        return completedTodos;
+
+      case StatusFilterType.ACTIVE:
+        return activeTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
+  const handleErrorClose = () => {
+    setError(null);
+  };
+
+  const addTodo = useCallback(async (title: string) => {
+    try {
+      const newTodo = {
+        userId: USER_ID,
+        completed: false,
+        title,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos((prevTodos) => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  }, []);
+
+  const removeTodo = useCallback(async (todoId: number) => {
+    try {
+      setLoadingTodoIds(prevTodoId => [...prevTodoId, todoId]);
+      await deleteTodo(todoId);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds([0]);
```

Comment:
What would be if you have 2 the same actions in one moment, but the second action will be slow for some time?
Remove only existing todo id from `LoadingTodoIds`